DOC/CI: include api docs on travis

jorisvandenbossche opened the pull request:
Continued in #12002

---

Closes #6100
Closes #3800

Just to check how this takes on travis

jorisvandenbossche committed:
DOC/CI: include api docs on travis

shoyer commented on doc/source/api.rst:
```diff
@@ -978,7 +978,7 @@ specific plotting methods of the form ``DataFrame.plot.<kind>``.
 .. autosummary::
    :toctree: generated/
-   :template: autosummary/accessor_plot.rst
+   :template: autosummary/accessor_callable.rst
```
oops! looks like this snuck through before

jreback commented:
might simply want to create a doc-build (e.g. this issue https://github.com/pydata/pandas/issues/3800)

jorisvandenbossche commented:
@jreback yes, I know, but I was first just experimenting a bit to get it working (not very familiar with travis/bash syntax details). Once I have something working the idea was indeed to put it in a separate travis build.

The problem I have now, is that for some reason `make.py` does not output its progress to the standard output of travis (which is the case if you run it locally in a terminal). And then, because of this, travis thinks nothing is happening anymore and aborts the build .. 
Do you know a solution for this?

jreback commented:
http://docs.travis-ci.com/user/build-timeouts/

might work

jreback commented:
is make.py writing to stderr? then could just redirect

maybe it's buffering?

jreback commented:
https://coderwall.com/p/wws2uq/have-travis-ci-test-your-sphinx-docs

?

jreback commented:
https://github.com/Syntaf/travis-sphinx

jorisvandenbossche commented:
`make.py` calls `sphinx-build`, which AFAIU outputs to both stdout as stderr? And also, I would think travis outputs both stdout as stderr in the logs? But it is not really clear to me how it is specified the output of the doc build is logged in a file that is then later shown, and why for tests it is outputted directly.

The `travis_wait` is probably not a solution, as this just raises the timeout from 10 to 20 minutes, but I think the doc build will even take longer.

jorisvandenbossche commented:
aha, hadn't notices that there is also a separate `run_build_docs.sh` script next to `build_docs.sh`, and that is of course capturing that output ..

jorisvandenbossche committed:
temp: let other fail

jorisvandenbossche committed:
DOC: hack to numpydoc to avoid warnings for Categorical (not including members)

The Categorical docstring in api.rst uses a seperate template to not
include an autosummary table with toctrees of all members. But
numpydoc still includes a list by default (which will gives thousands of
warnings). This hack ensures that for Categorical class docstring this
list is not included.

jorisvandenbossche committed:
DOC: add str accessor docstring pages to api.rst to avoid warning

jorisvandenbossche committed:
DOC: hack to numpydoc to include attributes that are None (GH6100)

See GH6100. Some attributes of our classes are None, even on the
class objects themselves (so not on instances). By default, numpydoc
does not include them in the autosummary table but in a separate
plain table. But, our class.rst template does include them in the
toctree autosummary, which creates docstring pages for these. This
results in warnings "WARNING: document isn't included in any toctree".
This hack to numpydoc does include them in the autosummary in the class
docstring, removing this warning.

jorisvandenbossche commented:
@jreback With the last commit the full docs build OK, although it took 37 minutes. So I put it in a separate travis build, and include the full api docs (there is still a temporary commit to let the others fail).

Can you have a look at the travis ci changes? (it also included somes hacks to numpydoc to prevent some warnings, they are not very nice hacks, but the warnings had to be eliminated to get it running on travis (otherwise the log was too big ..)).

jreback commented on .travis.yml:
```diff
@@ -18,6 +18,7 @@ git:
 
 matrix:
     fast_finish: true
+    fail_fast: true
```
I actually don't think we should have this and leave it false. That way you can see the full gamut of errors (of course for your testsing ok)

jorisvandenbossche replied:
yes, this one is from the temporary commit just for testing, will remove it

jreback commented on ci/before_install.sh:
```diff
@@ -11,3 +11,7 @@ echo "inside $0"
+
+if [ x"$DOC_BUILD" == x"" ]; then
+    exit 1
+fi
```
you nee to put this before the 'true' above. IOW, the script should *always* return true.
(I think this was for testing only?)

jorisvandenbossche replied:
yes, this was only for the testing from the temp commit (just so the others will all fail before installing pandas, so the doc build did not need to wait on all the others)

jreback commented on ci/script.sh:
```diff
@@ -4,6 +4,10 @@ echo "inside $0"
 source activate pandas
 
+if [ x"$DOC_BUILD" != x"" ]; then
+    exit 0
```
for testing?

jorisvandenbossche replied:
this one is on purpose and not for testing, so that the doc build does not run the tests (only the docs)

jreback commented:
@jorisvandenbossche all looks cool to me then. I know the doc-build takes a while but so what its in the  lower section.

note - the builds run in the order that they are defined in the top section (no matter if they are in th ebottom section or not). So you prob want to have the doc-build run last? (I think you have it that way). That way other workers can finish the other jobs. we can have 5 simultaneously i think.

jorisvandenbossche committed:
DOC: parallel build

jorisvandenbossche commented:
The problem with this approach is that the time it takes to build on Travis seems rather variable (maybe depending on its load?). In any case, sometimes it succeeded to finish it in ca 45 min, but other times it goes up to 50min and then travis aborts it. 
But now trying if parallel build speeds it up

jreback commented:
@jorisvandenbossche closing, but reopen if updating

in theory this looks like a great idea....but yeh it takes quite some time to build the API docs

jorisvandenbossche commented:
Continued in #12002
